src/modules/emote_menu/components/VirtualizedList.jsx

```diff
@@ -0,0 +1,111 @@
+import classNames from 'classnames';
+import React, {useCallback, useEffect, useMemo, useRef, useState} from 'react';
+import styles from '../styles/list.module.css';
+
+function VirtualizedList(
+  {className, totalRows, rowHeight, renderRow, windowHeight, stickyRows = [], onHeaderChange = () => {}},
+  ref
+) {
+  const listHeight = useMemo(() => rowHeight * totalRows, [totalRows, rowHeight]);
+
+  const [data, setData] = useState({
+    header: null,
+    rows: [],
+  });
+
+  useEffect(() => {
+    onHeaderChange(data.header);
+  }, [data.header]);
+
+  const wrapperRef = ref || useRef(null);
+
+  const isInViewport = useCallback(() => {
+    const {scrollTop} = wrapperRef.current;
+    const scrollBottom = scrollTop + windowHeight;
+
+    const startIndex = Math.floor(scrollTop / rowHeight);
+    const endIndex = Math.min(totalRows - 1, Math.floor(scrollBottom / rowHeight));
+
+    const rowsVisible = [];
+
+    for (let i = startIndex; i <= endIndex; i++) {
+      rowsVisible.push(i);
+    }
+
+    let current = 0;
+    let next = 0;
+
+    next = stickyRows.find((a) => {
+      if (a > startIndex) {
+        return true;
+      }
+
+      current = a;
+      return false;
+    });
+
+    const isStuck = startIndex < next - 1;
+
+    // current: the header that's closest to the top
+    // position: once another header hits the current header it'll become absolute and be "pushed" up
+    // rows: the rows that are visible in the viewport
+
+    setData({
+      header: isStuck
+        ? {
+            current,
+            top: 0,
+            position: 'sticky',
+          }
+        : {
+            current,
+            top: (next - 1) * rowHeight,
```

Comment:
the header logic is pretty buggy, so you will need to iron this out. it looks like it doesn't update until scroll stop. the animations also only work one direction (scrolling down).